File: backend/app/api/agents.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.core.security import require_role
from app.db.database import get_db
from app.models.agent import Agent
from app.models.user import User, UserRole
from app.schemas.agent import AgentCreate, AgentResponse
from app.services.assignment_service import assign_order

router = APIRouter()


@router.post(
    "/",
    response_model=AgentResponse,
    status_code=201,
)
def create_agent(
    data: AgentCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(
        require_role(UserRole.ADMIN)
    ),
):
    agent = Agent(
        name=data.name,
        phone=data.phone,
        active=True,
        available=True,
    )

    db.add(agent)
    db.commit()
    db.refresh(agent)

    return agent


@router.get("/")
def list_agents(
    db: Session = Depends(get_db),
):
    return db.query(Agent).all()


@router.post("/assign/{tracking_number}")
def assign_order_endpoint(
    tracking_number: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(
        require_role(UserRole.ADMIN)
    ),
):
    try:
        return assign_order(
            db,
            tracking_number,
        )

    except ValueError as e:
        raise HTTPException(
            status_code=400,
            detail=str(e),
        )

File: frontend/src/api/agents.ts
import { apiClient } from './client';
import { AgentAssignmentResponse, AgentCreate, AgentResponse } from '../types';

export const agentsApi = {
  /**
   * List all agents (public/authenticated)
   */
  getAgents: async (): Promise<AgentResponse[]> => {
    const res = await apiClient.get<AgentResponse[]>('/agents/');
    return res.data;
  },

  /**
   * Create a new agent (admin only)
   */
  createAgent: async (data: AgentCreate): Promise<AgentResponse> => {
    const res = await apiClient.post<AgentResponse>('/agents/', data);
    return res.data;
  },

  /**
   * Assign an available agent to an order (admin only)
   */
  assignOrder: async (trackingNumber: string): Promise<AgentAssignmentResponse> => {
    const res = await apiClient.post<AgentAssignmentResponse>(
      `/agents/assign/${encodeURIComponent(trackingNumber)}`
    );
    return res.data;
  },
};

